### DISCUSSION TITLE
Security: How is our extension supposed to handle processing of paths outside of a trusted workspace in regards to the trusted workspace feature?

### DISCUSSION THREAD

**Discussion by sean-mcmanus** [microsoft/vscode-discussions#2378] [Extension Development QnA] [answered]
If a user sets one of our extension's settings (e.g. includePath) to include the root folder (for example) or if their workspace has a symlink to the root, does that mean the user is granting our extension trust to all folders under the root or should our extension not be trusting external folders that are not on the trusted workspaces list? i.e. can we run executables from there, show users links contained in those files, etc. VS Code itself doesn't trust the paths if the user opens a file or workspace directly.

Should the VS Code trusted workspace UI be updated to warn users that any folders added to extensions settings are transitively trusted? I don't know how users are supposed to know this.

I couldn't find the answer at https://code.visualstudio.com/api/extension-guides/workspace-trust .

**Comment by alefragnani** [comment]
Hi @sean-mcmanus ,

Workspace Trust are primarily focused on _trusting_ the content of the folder/workspace you are opening on VS Code, with the intention to protect the user from malicious files that may contains inside an unknown folder/workspace, either executables or other files that may harm user's system when some external executable uses it.

On the other hand, trusting on _settings_ is a bit of a mix, and not entirely related to Workspace Trust feature. But, the Workspace Trust API allows you to make some of your extensions setting aware of a trusted workspace, and _limit_ its functionality if you don't trust on some folder/workspace, using the `restrictedConfiguration` setting (https://code.visualstudio.com/api/extension-guides/workspace-trust#configurations-settings).

Specifically about a _symlink_ contained inside a folder/workspace a user is opening, it end up being a user's decision, which may or may not, trust on that folder/workspace. If the user _trusts_, it means that the symlink will also be trusted. 

Should VS Code improve the UI to better explain situations like _symlink_ or malicious files references in any setting, maybe. If that's the case, I would suggest you to open a feature request in VS Code repo https://github.com/microsoft/vscode/issues

Hope this helps

**Reply by sean-mcmanus** [reply]
I'm not sure I understand. If a user opens folder /repos/trusted and but has setting "${workspaceFolder}/.." that is scanned recursively should our extension trust /repos/untrusted, even if VS Code itself counts it as untrusted and blocks files from being opened?

In regards to a symlink, a user (or a tool they run) might have created the symlink to /Users and trust the symlink itself (it points to a known location that isn't in itself malicious), but the user might not understand that a particular extension will treat that as if the user trusts /Users/maliciousUser as well.

Since this scenario isn't documented, I'm thinking it might be out-of-scope for what the trusted workspace feature was designed to handle.

**Reply by alefragnani** [reply, marked as answer]
I think your scenario is beyond what Workspace Trust is proposed to. Workspace Trust does not define a set of folders that are _safe to be used anywhere in VS Code, including settings_, but only a set of folders are _safe to be opened by VS Code_ itself. 

With that in mind, it does not matter (for Workspace Trust API) if the user puts an _unsafe path_ on any extension setting. It only matters if the user opened a trusted folder or not. So, in your case, if the user trusts `repos/trusted` and you have a setting on your extension defined as `${workspaceFolder}/` which points to `repos/`, VS Code won't know about the meaning of your setting, and besides not loading files from `/repos/untrested` (because only `repos/trusted` is trusted), it won't block _your extension_ from loading files from there. 

But, I understand your point, and yes, maybe the API could be improved so extension authors could also know the trusted folders, which does not exists today.
